File: src/number_literal.cpp
/*
 * Copyright (c) 2026-present Raphael Amorim
 *
 * This file is part of jam.
 * Licensed under the Apache License, Version 2.0 with LLVM Exceptions.
 */

#include "number_literal.h"

#include <cstdlib>  // strtod
#include <string>

namespace {

NumberResult fail(NumberErrorKind kind, std::size_t pos,
                  NumberBase base = NumberBase::Decimal) {
	NumberResult r;
	r.kind = NumberResultKind::Failure;
	r.failure.kind = kind;
	r.failure.pos = pos;
	r.failure.base = base;
	return r;
}

NumberResult okInt(uint64_t value, NumberBase base) {
	NumberResult r;
	r.kind = NumberResultKind::Int;
	r.intValue = value;
	r.base = base;
	return r;
}

NumberResult okBigInt(NumberBase base) {
	NumberResult r;
	r.kind = NumberResultKind::BigInt;
	r.base = base;
	return r;
}

NumberResult okFloat(NumberBase base, double value) {
	NumberResult r;
	r.kind = NumberResultKind::Float;
	r.base = base;
	r.floatValue = value;
	return r;
}

}  // namespace

NumberResult parseNumberLiteral(std::string_view bytes) {
	std::size_t i = 0;
	uint8_t baseValue = 10;
	NumberBase base = NumberBase::Decimal;

	if (bytes.size() >= 2 && bytes[0] == '0') {
		switch (bytes[1]) {
		case 'b':
			base = NumberBase::Binary;
			baseValue = 2;
			i = 2;
			break;
		case 'o':
			base = NumberBase::Octal;
			baseValue = 8;
			i = 2;
			break;
		case 'x':
			base = NumberBase::Hex;
			baseValue = 16;
			i = 2;
			break;
		case 'B':
		case 'O':
		case 'X':
			// Reject uppercase prefix outright for stylistic uniformity;
			// callers must spell `0x` / `0b` / `0o` in lowercase.
			return fail(NumberErrorKind::UpperCaseBase, 1);
		case '.':
		case 'e':
		case 'E':
			// Float starting with "0." / "0e..." — fall through to the
			// decimal scanner.
			break;
		default:
			// Any other leading-zero form ("07", "0a", etc.) is an error
			// to avoid silent C-style octal interpretation.
			return fail(NumberErrorKind::LeadingZero, 0);
		}
	}

	// `0x` / `0b` / `0o` with no following digit is an error.
	if (bytes.size() == 2 && baseValue != 10) {
		return fail(NumberErrorKind::DigitAfterBase, 2);
	}

	uint64_t x = 0;
	bool overflow = false;
	bool underscore = false;
	bool period = false;
	uint8_t special = 0;
	bool exponent = false;
	bool isFloat = false;

	while (i < bytes.size()) {
		char c = bytes[i];

		// Digit separator. Cannot appear immediately after a base prefix
		// or after a special character (period, sign, exponent letter),
		// and cannot repeat.
		if (c == '_') {
			if (i == 2 && baseValue != 10) {
				return fail(NumberErrorKind::InvalidUnderscoreAfterSpecial, i);
			}
			if (special != 0) {
				return fail(NumberErrorKind::InvalidUnderscoreAfterSpecial, i);
			}
			if (underscore) {
				return fail(NumberErrorKind::RepeatedUnderscore, i);
			}
			underscore = true;
			i++;
			continue;
		}

		// Decimal exponent marker. Only valid in decimal — in hex
		// literals `e` / `E` are hex digits, never exponent letters.
		if ((c == 'e' || c == 'E') && baseValue == 10) {
			isFloat = true;
			if (exponent) {
				return fail(NumberErrorKind::DuplicateExponent, i);
			}
			if (underscore) {
				return fail(NumberErrorKind::ExponentAfterUnderscore, i);
			}
			special = static_cast<uint8_t>(c);
			exponent = true;
			i++;
			continue;
		}

		// Hex / binary float exponent marker (always binary-base
		// exponent in IEEE-754 hex floats).
		if ((c == 'p' || c == 'P') && baseValue == 16) {
			isFloat = true;
			if (exponent) {
				return fail(NumberErrorKind::DuplicateExponent, i);
			}
			if (underscore) {
				return fail(NumberErrorKind::ExponentAfterUnderscore, i);
			}
			special = static_cast<uint8_t>(c);
			exponent = true;
			i++;
			continue;
		}

		// Decimal point. Only meaningful in decimal or hex; binary /
		// octal floats are not a thing.
		if (c == '.') {
			isFloat = true;
			if (baseValue != 10 && baseValue != 16) {
				return fail(NumberErrorKind::InvalidFloatBase, 2);
			}
			if (period) { return fail(NumberErrorKind::DuplicatePeriod, i); }
			period = true;
			if (underscore) {
				return fail(NumberErrorKind::SpecialAfterUnderscore, i);
			}
			special = '.';
			i++;
			continue;
		}

		// Sign on exponent. Only valid immediately after p/P/e/E.
		if (c == '+' || c == '-') {
			if (special != 'p' && special != 'P' && special != 'e' &&
			    special != 'E') {
				return fail(NumberErrorKind::InvalidExponentSign, i);
			}
			special = static_cast<uint8_t>(c);
			i++;
			continue;
		}

		// At this point `c` should be a digit in the current base. Map
		// it to its numeric value.
		uint8_t digit;
		if (c >= '0' && c <= '9') {
			digit = static_cast<uint8_t>(c - '0');
		} else if (c >= 'A' && c <= 'Z') {
			digit = static_cast<uint8_t>(c - 'A' + 10);
		} else if (c >= 'a' && c <= 'z') {
			digit = static_cast<uint8_t>(c - 'a' + 10);
		} else {
			return fail(NumberErrorKind::InvalidCharacter, i);
		}

		if (digit >= baseValue) {
			return fail(NumberErrorKind::InvalidDigit, i, base);
		}
		if (exponent && digit >= 10) {
			return fail(NumberErrorKind::InvalidDigitExponent, i);
		}

		underscore = false;
		special = 0;

		if (!isFloat) {
			// Accumulate value, watching for overflow.
			if (x != 0) {
				uint64_t mul = x * baseValue;
				if (mul / baseValue != x) overflow = true;
				x = mul;
			}
			uint64_t add = x + digit;
			if (add < x) overflow = true;
			x = add;
		}

		i++;
	}

	if (underscore) {
		return fail(NumberErrorKind::TrailingUnderscore, bytes.size() - 1);
	}
	if (special != 0) {
		return fail(NumberErrorKind::TrailingSpecial, bytes.size() - 1);
	}

	if (isFloat) {
		// rule:
		// Decimal and hex floats both parse cleanly with the standard C
		// `strtod` (it accepts `0x1.8p4` for hex). Underscores have been
		// validated already; strip them before strtod so it sees a
		// pristine numeric form.
		std::string clean;
		clean.reserve(bytes.size());
		for (char c : bytes) {
			if (c != '_') clean.push_back(c);
		}
		const char *cstr = clean.c_str();
		char *end = nullptr;
		double v = std::strtod(cstr, &end);
		// `end` should land at the end of the string; if not, we caught
		// an unparseable character above and shouldn't have reached
		// here, but guard for safety.
		(void)end;
		return okFloat(base, v);
	}
	if (overflow) return okBigInt(base);
	return okInt(x, base);
}

const char *numberErrorMessage(NumberErrorKind kind) {
	switch (kind) {
	case NumberErrorKind::LeadingZero:
		return "leading zero is not allowed (use `0o` for octal)";
	case NumberErrorKind::DigitAfterBase:
		return "expected digit after base prefix";
	case NumberErrorKind::UpperCaseBase:
		return "base prefix must be lowercase (`0x`, `0b`, `0o`)";
	case NumberErrorKind::InvalidFloatBase:
		return "float literals are only allowed in decimal or hex";
	case NumberErrorKind::RepeatedUnderscore:
		return "consecutive underscores in numeric literal";
	case NumberErrorKind::InvalidUnderscoreAfterSpecial:
		return "underscore not allowed immediately after `0x`/`0b`/`0o` "
		       "or after `.`/`+`/`-`/`e`/`E`/`p`/`P`";
	case NumberErrorKind::InvalidDigit:
		return "invalid digit for the literal's base";
	case NumberErrorKind::InvalidDigitExponent:
		return "non-decimal digit in exponent";
	case NumberErrorKind::DuplicatePeriod:
		return "more than one `.` in numeric literal";
	case NumberErrorKind::DuplicateExponent:
		return "more than one exponent in numeric literal";
	case NumberErrorKind::InvalidHexExponent:
		return "hex literal cannot use `e`/`E` exponent (use `p`/`P`)";
	case NumberErrorKind::ExponentAfterUnderscore:
		return "exponent letter cannot follow `_`";
	case NumberErrorKind::SpecialAfterUnderscore:
		return "`.`/`+`/`-` cannot follow `_`";
	case NumberErrorKind::TrailingSpecial:
		return "numeric literal ends with `.`/`+`/`-`/`e`/`p`";
	case NumberErrorKind::TrailingUnderscore:
		return "numeric literal ends with `_`";
	case NumberErrorKind::InvalidCharacter:
		return "character not allowed in numeric literal";
	case NumberErrorKind::InvalidExponentSign:
		return "`+`/`-` is only valid immediately after `e`/`E`/`p`/`P`";
	case NumberErrorKind::IntegerTooLarge:
		return "integer literal exceeds u64 range";
	}
	return "unknown numeric-literal error";
}

File: crates/jam-syntax/src/number_literal.rs
/*
 * Copyright (c) 2026-present Raphael Amorim
 *
 * This file is part of jam.
 * Licensed under the Apache License, Version 2.0 with LLVM Exceptions.
 */

//! Numeric-literal validation and value extraction. The lexer produces a
//! permissive `NUMBER` token; here its bytes are validated (base prefix,
//! underscore placement, digit ranges, exponent rules) and converted to a
//! value. Overflowing integers become `BigInt`; floats parse via
//! [`float128::parse_decimal_float`] (correctly rounded, decimal and hex).

use float128::{ParsedFloat, parse_decimal_float, quad_to_target_as_double};

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
#[repr(u8)]
pub enum NumberBase {
    Decimal = 10,
    Hex = 16,
    Binary = 2,
    Octal = 8,
}

/// Per-error variants with the byte offset into the lexeme where the problem
/// was found.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum NumberErrorKind {
    LeadingZero,
    DigitAfterBase,
    UpperCaseBase,
    InvalidFloatBase,
    RepeatedUnderscore,
    InvalidUnderscoreAfterSpecial,
    InvalidDigit,
    InvalidDigitExponent,
    DuplicatePeriod,
    DuplicateExponent,
    InvalidHexExponent,
    ExponentAfterUnderscore,
    SpecialAfterUnderscore,
    TrailingSpecial,
    TrailingUnderscore,
    InvalidCharacter,
    InvalidExponentSign,
    IntegerTooLarge,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct NumberError {
    pub kind: NumberErrorKind,
    pub pos: usize,
    /// Meaningful for `InvalidDigit`; otherwise `Decimal`.
    pub base: NumberBase,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum NumberResultKind {
    Int,
    BigInt,
    Float,
    Failure,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NumberResult {
    pub kind: NumberResultKind,
    pub int_value: u64,
    pub float_value: f64,
    pub base: NumberBase,
    pub failure: Option<NumberError>,
}

impl NumberResult {
    fn fail(kind: NumberErrorKind, pos: usize, base: NumberBase) -> NumberResult {
        NumberResult {
            kind: NumberResultKind::Failure,
            int_value: 0,
            float_value: 0.0,
            base: NumberBase::Decimal,
            failure: Some(NumberError { kind, pos, base }),
        }
    }
    fn ok_int(value: u64, base: NumberBase) -> NumberResult {
        NumberResult {
            kind: NumberResultKind::Int,
            int_value: value,
            float_value: 0.0,
            base,
            failure: None,
        }
    }
    fn ok_big_int(base: NumberBase) -> NumberResult {
        NumberResult {
            kind: NumberResultKind::BigInt,
            int_value: 0,
            float_value: 0.0,
            base,
            failure: None,
        }
    }
    fn ok_float(base: NumberBase, value: f64) -> NumberResult {
        NumberResult {
            kind: NumberResultKind::Float,
            int_value: 0,
            float_value: value,
            base,
            failure: None,
        }
    }
}

/// Validate and parse a numeric literal lexeme (raw token bytes).
pub fn parse_number_literal(bytes: &[u8]) -> NumberResult {
    let mut i = 0usize;
    let mut base_value: u8 = 10;
    let mut base = NumberBase::Decimal;

    if bytes.len() >= 2 && bytes[0] == b'0' {
        match bytes[1] {
            b'b' => {
                base = NumberBase::Binary;
                base_value = 2;
                i = 2;
            }
            b'o' => {
                base = NumberBase::Octal;
                base_value = 8;
                i = 2;
            }
            b'x' => {
                base = NumberBase::Hex;
                base_value = 16;
                i = 2;
            }
            b'B' | b'O' | b'X' => {
                return NumberResult::fail(NumberErrorKind::UpperCaseBase, 1, NumberBase::Decimal);
            }
            // "0." / "0e" / "0E" — decimal float; fall through to the scanner.
            b'.' | b'e' | b'E' => {}
            _ => {
                return NumberResult::fail(NumberErrorKind::LeadingZero, 0, NumberBase::Decimal);
            }
        }
    }

    if bytes.len() == 2 && base_value != 10 {
        return NumberResult::fail(NumberErrorKind::DigitAfterBase, 2, NumberBase::Decimal);
    }

    let mut x: u64 = 0;
    let mut overflow = false;
    let mut underscore = false;
    let mut period = false;
    let mut special: u8 = 0;
    let mut exponent = false;
    let mut is_float = false;

    while i < bytes.len() {
        let c = bytes[i];

        if c == b'_' {
            if i == 2 && base_value != 10 {
                return NumberResult::fail(
                    NumberErrorKind::InvalidUnderscoreAfterSpecial,
                    i,
                    NumberBase::Decimal,
                );
            }
            if special != 0 {
                return NumberResult::fail(
                    NumberErrorKind::InvalidUnderscoreAfterSpecial,
                    i,
                    NumberBase::Decimal,
                );
            }
            if underscore {
                return NumberResult::fail(
                    NumberErrorKind::RepeatedUnderscore,
                    i,
                    NumberBase::Decimal,
                );
            }
            underscore = true;
            i += 1;
            continue;
        }

        if (c == b'e' || c == b'E') && base_value == 10 {
            is_float = true;
            if exponent {
                return NumberResult::fail(
                    NumberErrorKind::DuplicateExponent,
                    i,
                    NumberBase::Decimal,
                );
            }
            if underscore {
                return NumberResult::fail(
                    NumberErrorKind::ExponentAfterUnderscore,
                    i,
                    NumberBase::Decimal,
                );
            }
            special = c;
            exponent = true;
            i += 1;
            continue;
        }

        if (c == b'p' || c == b'P') && base_value == 16 {
            is_float = true;
            if exponent {
                return NumberResult::fail(
                    NumberErrorKind::DuplicateExponent,
                    i,
                    NumberBase::Decimal,
                );
            }
            if underscore {
                return NumberResult::fail(
                    NumberErrorKind::ExponentAfterUnderscore,
                    i,
                    NumberBase::Decimal,
                );
            }
            special = c;
            exponent = true;
            i += 1;
            continue;
        }

        if c == b'.' {
            is_float = true;
            if base_value != 10 && base_value != 16 {
                return NumberResult::fail(
                    NumberErrorKind::InvalidFloatBase,
                    2,
                    NumberBase::Decimal,
                );
            }
            if period {
                return NumberResult::fail(
                    NumberErrorKind::DuplicatePeriod,
                    i,
                    NumberBase::Decimal,
                );
            }
            period = true;
            if underscore {
                return NumberResult::fail(
                    NumberErrorKind::SpecialAfterUnderscore,
                    i,
                    NumberBase::Decimal,
                );
            }
            special = b'.';
            i += 1;
            continue;
        }

        if c == b'+' || c == b'-' {
            if special != b'p' && special != b'P' && special != b'e' && special != b'E' {
                return NumberResult::fail(
                    NumberErrorKind::InvalidExponentSign,
                    i,
                    NumberBase::Decimal,
                );
            }
            special = c;
            i += 1;
            continue;
        }

        // Map `c` to a digit value in the current base.
        let digit: u8 = if c.is_ascii_digit() {
            c - b'0'
        } else if c.is_ascii_uppercase() {
            c - b'A' + 10
        } else if c.is_ascii_lowercase() {
            c - b'a' + 10
        } else {
            return NumberResult::fail(NumberErrorKind::InvalidCharacter, i, NumberBase::Decimal);
        };

        if digit >= base_value {
            return NumberResult::fail(NumberErrorKind::InvalidDigit, i, base);
        }
        if exponent && digit >= 10 {
            return NumberResult::fail(
                NumberErrorKind::InvalidDigitExponent,
                i,
                NumberBase::Decimal,
            );
        }

        underscore = false;
        special = 0;

        if !is_float {
            // Accumulate with wrapping arithmetic, tracking overflow.
            if x != 0 {
                let mul = x.wrapping_mul(base_value as u64);
                if mul / base_value as u64 != x {
                    overflow = true;
                }
                x = mul;
            }
            let add = x.wrapping_add(digit as u64);
            if add < x {
                overflow = true;
            }
            x = add;
        }

        i += 1;
    }

    if underscore {
        return NumberResult::fail(
            NumberErrorKind::TrailingUnderscore,
            bytes.len() - 1,
            NumberBase::Decimal,
        );
    }
    if special != 0 {
        return NumberResult::fail(
            NumberErrorKind::TrailingSpecial,
            bytes.len() - 1,
            NumberBase::Decimal,
        );
    }

    if is_float {
        // Underscores already validated — strip them and parse the pristine
        // numeric form. `parse_decimal_float` handles decimal AND hex floats
        // (`0x1.8p4`) with one correctly-rounded step.
        let clean: Vec<u8> = bytes.iter().copied().filter(|&c| c != b'_').collect();
        let clean_str = std::str::from_utf8(&clean).unwrap_or("");
        let v = match parse_decimal_float(clean_str) {
            ParsedFloat::Fits64(bits) => f64::from_bits(bits),
            ParsedFloat::Quad(q) => quad_to_target_as_double(&q, false),
        };
        return NumberResult::ok_float(base, v);
    }
    if overflow {
        return NumberResult::ok_big_int(base);
    }
    NumberResult::ok_int(x, base)
}

/// Human-readable message for a numeric-literal error.
pub fn number_error_message(kind: NumberErrorKind) -> &'static str {
    use NumberErrorKind::*;
    match kind {
        LeadingZero => "leading zero is not allowed (use `0o` for octal)",
        DigitAfterBase => "expected digit after base prefix",
        UpperCaseBase => "base prefix must be lowercase (`0x`, `0b`, `0o`)",
        InvalidFloatBase => "float literals are only allowed in decimal or hex",
        RepeatedUnderscore => "consecutive underscores in numeric literal",
        InvalidUnderscoreAfterSpecial => {
            "underscore not allowed immediately after `0x`/`0b`/`0o` or after `.`/`+`/`-`/`e`/`E`/`p`/`P`"
        }
        InvalidDigit => "invalid digit for the literal's base",
        InvalidDigitExponent => "non-decimal digit in exponent",
        DuplicatePeriod => "more than one `.` in numeric literal",
        DuplicateExponent => "more than one exponent in numeric literal",
        InvalidHexExponent => "hex literal cannot use `e`/`E` exponent (use `p`/`P`)",
        ExponentAfterUnderscore => "exponent letter cannot follow `_`",
        SpecialAfterUnderscore => "`.`/`+`/`-` cannot follow `_`",
        TrailingSpecial => "numeric literal ends with `.`/`+`/`-`/`e`/`p`",
        TrailingUnderscore => "numeric literal ends with `_`",
        InvalidCharacter => "character not allowed in numeric literal",
        InvalidExponentSign => "`+`/`-` is only valid immediately after `e`/`E`/`p`/`P`",
        IntegerTooLarge => "integer literal exceeds u64 range",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(s: &str) -> NumberResult {
        parse_number_literal(s.as_bytes())
    }

    #[test]
    fn integers_by_base() {
        assert_eq!(parse("42").kind, NumberResultKind::Int);
        assert_eq!(parse("42").int_value, 42);
        assert_eq!(parse("0x800").int_value, 0x800); // test_array_type_sizes
        assert_eq!(parse("0x800").base, NumberBase::Hex);
        assert_eq!(parse("0b1010").int_value, 0b1010);
        assert_eq!(parse("0o17").int_value, 0o17);
        assert_eq!(parse("1_000_000").int_value, 1_000_000);
    }

    #[test]
    fn overflow_is_bigint() {
        // 2^64 overflows u64.
        assert_eq!(parse("18446744073709551616").kind, NumberResultKind::BigInt);
        assert_eq!(parse("18446744073709551615").int_value, u64::MAX);
    }

    #[test]
    #[allow(clippy::approx_constant)] // `3.14` is the parse input under test, not an approximation of PI
    fn floats_decimal_and_hex() {
        let d = parse("3.14");
        assert_eq!(d.kind, NumberResultKind::Float);
        assert_eq!(d.float_value, 3.14_f64);
        assert_eq!(d.base, NumberBase::Decimal);

        let e = parse("1.5e-3");
        assert_eq!(e.float_value, 1.5e-3_f64);

        let h = parse("0x1.8p1");
        assert_eq!(h.kind, NumberResultKind::Float);
        assert_eq!(h.base, NumberBase::Hex);
        assert_eq!(h.float_value, 3.0_f64); // 1.5 * 2^1
    }

    #[test]
    fn float_matches_rust_parse() {
        for s in [
            "0.1",
            "3.141592653589793",
            "1e100",
            "2.5",
            "0.30000000000000004",
        ] {
            assert_eq!(parse(s).float_value, s.parse::<f64>().unwrap(), "{s}");
        }
    }

    #[test]
    fn rejects() {
        assert_eq!(
            parse("07").failure.unwrap().kind,
            NumberErrorKind::LeadingZero
        );
        assert_eq!(
            parse("0X42").failure.unwrap().kind,
            NumberErrorKind::UpperCaseBase
        );
        assert_eq!(
            parse("0x").failure.unwrap().kind,
            NumberErrorKind::DigitAfterBase
        );
        assert_eq!(
            parse("1__0").failure.unwrap().kind,
            NumberErrorKind::RepeatedUnderscore
        );
        assert_eq!(
            parse("1.0.5").failure.unwrap().kind,
            NumberErrorKind::DuplicatePeriod
        );
        assert_eq!(
            parse("0b2").failure.unwrap().kind,
            NumberErrorKind::InvalidDigit
        );
        assert_eq!(
            parse("1_").failure.unwrap().kind,
            NumberErrorKind::TrailingUnderscore
        );
        // base reported on InvalidDigit
        assert_eq!(parse("0b2").failure.unwrap().base, NumberBase::Binary);
    }
}
